pkg/build/grafana.go

```diff
@@ -0,0 +1,136 @@
+// Copyright Istio Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+
+	"istio.io/release-builder/pkg/model"
+	"istio.io/release-builder/pkg/util"
+)
+
+// Grafana packages Istio dashboards in a form that is ready to be published to grafana.com
+func Grafana(manifest model.Manifest) error {
+	if err := util.CopyDir(
+		path.Join(manifest.RepoDir("istio"), "manifests/istio-telemetry/grafana/dashboards"),
+		path.Join(manifest.WorkDir(), "grafana"),
+	); err != nil {
+		return err
+	}
+	dashboards, err := ioutil.ReadDir(path.Join(manifest.WorkDir(), "grafana"))
+	if err != nil {
+		return fmt.Errorf("failed to read dashboards: %v", err)
+	}
+	for _, dashboard := range dashboards {
+		if err := externalizeDashboard(manifest.Version, path.Join(path.Join(manifest.WorkDir(), "grafana", dashboard.Name()))); err != nil {
+			return fmt.Errorf("failed to process dashboard %v: %v", dashboard.Name(), err)
+		}
+	}
+	if err := util.CopyDir(
+		path.Join(manifest.WorkDir(), "grafana"),
+		path.Join(manifest.OutDir(), "grafana"),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+// externalizeDashboard converts a grafana dashboard from the "internal" representation, which is used
+// in the charts, to the "external" representation. This is the form needed to publish to grafana.com
+// This has two fields added, __input and __requires, and the datasource is not hardcoded.
+func externalizeDashboard(version, file string) error {
+	original, err := ioutil.ReadFile(file)
+	if err != nil {
+		return fmt.Errorf("failed to read %v: %v", file, err)
+	}
+	var msg map[string]json.RawMessage
+	if err := json.Unmarshal(original, &msg); err != nil {
+		return fmt.Errorf("failed to unmarshall %v: %v", file, err)
+	}
+
+	var inputMsg json.RawMessage
+	if err := json.Unmarshal([]byte(`[
+    {
+      "name": "DS_PROMETHEUS",
+      "label": "Prometheus",
+      "description": "",
+      "type": "datasource",
+      "pluginId": "prometheus",
+      "pluginName": "Prometheus"
+    }
+  ]`), &inputMsg); err != nil {
+		return fmt.Errorf("failed to construct __input: %v", err)
+	}
+	msg["__input"] = inputMsg
+
+	var requiresMesh json.RawMessage
+	if err := json.Unmarshal([]byte(`[
+    {
+      "type": "grafana",
+      "id": "grafana",
+      "name": "Grafana",
+      "version": "6.4.3"
```

Comment:
should this be another param ?

Reply:
I basically took our dashboards, clicked export, and looked at what they added. All of them seemed to have this version, so I assume we use some panel or feature in that version? The only thing we will need to do is ensure that if we utilize a new feature we update this. I think the impact would just be that someone with an old grafana deployment may unexpectedly have some panel break though?

Any other ideas to improve this are welcome, this is a huge hack. I don't really want to try to spin up a whole grafana instance and extract this each build as well, seems a bit heavy-handed